source/common/router/scoped_rds.cc

```diff
@@ -0,0 +1,272 @@
+      auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
+          static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
+      thread_local_scoped_config->removeRoutingScope(scoped_route_name);
+    });
+  }
+
+  ConfigSubscriptionCommonBase::onConfigUpdate();
+  setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating scoped route(s) {}",
+                                     StringUtil::join(exception_msgs, ", ")));
+  }
+  stats_.config_reload_.inc();
+}
+
+ScopedRdsConfigProvider::ScopedRdsConfigProvider(
+    ScopedRdsConfigSubscriptionSharedPtr&& subscription,
+    Server::Configuration::FactoryContext& factory_context,
+    envoy::api::v2::core::ConfigSource rds_config_source,
+    const envoy::config::filter::network::http_connection_manager::v2::ScopedRoutes::
+        ScopeKeyBuilder& scope_key_builder)
+    : DeltaMutableConfigProviderBase(std::move(subscription), factory_context,
+                                     ConfigProvider::ApiType::Delta),
+      subscription_(static_cast<ScopedRdsConfigSubscription*>(
+          MutableConfigProviderCommonBase::subscription_.get())),
+      rds_config_source_(std::move(rds_config_source)) {
+  initialize([scope_key_builder](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    return std::make_shared<ThreadLocalScopedConfigImpl>(scope_key_builder);
+  });
+}
+
+ProtobufTypes::MessagePtr ScopedRoutesConfigProviderManager::dumpConfigs() const {
+  auto config_dump = std::make_unique<envoy::admin::v2alpha::ScopedRoutesConfigDump>();
+  for (const auto& element : configSubscriptions()) {
+    auto subscription = element.second.lock();
+    ASSERT(subscription);
+
+    if (subscription->configInfo()) {
+      auto* dynamic_config = config_dump->mutable_dynamic_scoped_route_configs()->Add();
+      dynamic_config->set_version_info(subscription->configInfo().value().last_config_version_);
+      const ScopedRdsConfigSubscription* typed_subscription =
+          static_cast<ScopedRdsConfigSubscription*>(subscription.get());
+      dynamic_config->set_name(typed_subscription->name());
+      const ScopedConfigManager::ScopedRouteMap& scoped_route_map =
+          typed_subscription->scopedRouteMap();
+      for (const auto& it : scoped_route_map) {
+        dynamic_config->mutable_scoped_route_configs()->Add()->MergeFrom(it.second->config_proto_);
+      }
+      TimestampUtil::systemClockToTimestamp(subscription->lastUpdated(),
+                                            *dynamic_config->mutable_last_updated());
+    }
+  }
+
+  for (const auto& provider : immutableConfigProviders(ConfigProviderInstanceType::Inline)) {
+    const auto protos_info =
+        provider->configProtoInfoVector<envoy::api::v2::ScopedRouteConfiguration>();
+    if (protos_info == absl::nullopt) {
```

Comment:
When would this be true? Maybe add a comment?

Reply:
Converted to `ASSERT()`.